### PR TITLE
TPCTransform: add an interface for the space charge correction

### PR DESCRIPTION
**Pull request by sgorbuno**
Fast TPC Transformation: 
- an interface for the initialisation of the space charge correction
- an example macro

### CODE REVIEW

File: Detectors/TPC/reconstruction/macro/getTPCTransformationExample.C [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+	  float x1g = x1*sliceInfo.cosAlpha - y1*sliceInfo.sinAlpha;
+	  float y1g = x1*sliceInfo.sinAlpha + y1*sliceInfo.cosAlpha;	  
+	  float z1g = z1;
+	  
+	  //cout<<x0<<" "<<y0<<" "<<z0<<" "<<x0g<<" "<<y0g<<" "<<z0g<<endl;
+	  //cout<<x1<<" "<<y1<<" "<<z1<<" "<<x1g<<" "<<y1g<<" "<<z1g<<endl;
+
+	  double xyz[3] = {x0g,y0g,z0g};
+	  double d[3] = {0,0,0};
+	  spaceChargeCorrection( xyz, d );
```
Comment (matthiasrichter) [inline Detectors/TPC/reconstruction/macro/getTPCTransformationExample.C]: I'm a bit puzzled by the fact that the correction function is called directly to calculate the parameters for the histogram. Shouldn't that be done in the transformation object?

File: Detectors/TPC/reconstruction/include/TPCReconstruction/TPCFastTransformHelperO2.h [hunk 2]
```diff
@@ -66,9 +72,10 @@ class TPCFastTransformHelperO2
   /// initialization
   void init();
 
-  static TPCFastTransformHelperO2* sInstance; // singleton instance
+  static TPCFastTransformHelperO2* sInstance; ///< singleton instance
   bool mIsInitialized;                        ///< initialization flag
-
+  void (*mSpaceChargeCorrection)( const double XYZ[3], double dXdYdZ[3] ); ///< pointer to an external correction method
```
[the remark this answers is not in the document]
Comment (sgorbuno) [reply Detectors/TPC/reconstruction/include/TPCReconstruction/TPCFastTransformHelperO2.h]: ok, fixed

File: Detectors/TPC/reconstruction/src/TPCFastTransformHelperO2.cxx [hunk 3]
```diff
@@ -44,7 +44,8 @@ TPCFastTransformHelperO2* TPCFastTransformHelperO2::instance()
 }
 
 TPCFastTransformHelperO2::TPCFastTransformHelperO2()
-  : mIsInitialized(0)
+  : mIsInitialized(0),
+    mSpaceChargeCorrection(nullptr)
```
[the remark this answers is not in the document]
Comment (sgorbuno) [reply Detectors/TPC/reconstruction/src/TPCFastTransformHelperO2.cxx]: ok, fixed

File: Detectors/TPC/reconstruction/src/TPCFastTransformHelperO2.cxx [hunk 4]
```diff
@@ -231,23 +234,83 @@ int TPCFastTransformHelperO2::updateCalibration(ali_tpc_common::tpc_fast_transfo
...
+
       const IrregularSpline2D3D& spline = distortion.getSpline(slice, row);
       float* data = distortion.getSplineDataNonConst(slice, row);
+
       for (int knot = 0; knot < spline.getNumberOfKnots(); knot++) {
-        data[3 * knot + 0] = 0.f;
-        data[3 * knot + 1] = 0.f;
-        data[3 * knot + 2] = 0.f;
+        
+	if( mSpaceChargeCorrection==nullptr ){
```
[the remark this answers is not in the document]
Comment (sgorbuno) [reply Detectors/TPC/reconstruction/src/TPCFastTransformHelperO2.cxx]: ok, fixed

File: Detectors/TPC/reconstruction/src/TPCFastTransformHelperO2.cxx [hunk 5]
```diff
@@ -63,6 +64,8 @@ std::unique_ptr<TPCFastTransform> TPCFastTransformHelperO2::create(Long_t TimeSt
 
   TPCFastTransform& fastTransform = *fastTransformPtr;
 
+  fastTransform.setApplyDistortionFlag(1);
```
[the remark this answers is not in the document]
Comment (sgorbuno) [reply Detectors/TPC/reconstruction/src/TPCFastTransformHelperO2.cxx]: ok, fixed

File: Detectors/TPC/reconstruction/include/TPCReconstruction/TPCFastTransformHelperO2.h [hunk 6]
```diff
@@ -66,9 +72,10 @@ class TPCFastTransformHelperO2
   /// initialization
   void init();
 
-  static TPCFastTransformHelperO2* sInstance; // singleton instance
+  static TPCFastTransformHelperO2* sInstance; ///< singleton instance
   bool mIsInitialized;                        ///< initialization flag
```
[the remark this answers is not in the document]
Comment (sgorbuno) [reply Detectors/TPC/reconstruction/include/TPCReconstruction/TPCFastTransformHelperO2.h]: ok, fixed

File: Detectors/TPC/reconstruction/include/TPCReconstruction/TPCFastTransformHelperO2.h [hunk 7]
```diff
@@ -52,6 +52,12 @@ class TPCFastTransformHelperO2
 
   /// _______________  Main functionality  ________________________
 
+  /// set an external space charge correction in the global coordinates
+
+  void setSpaceChargeCorrection( void (*spaceChargeCorrection)( const double XYZ[3], double dXdYdZ[3] ) ){
```
[the remark this answers is not in the document]
Comment (sgorbuno) [reply Detectors/TPC/reconstruction/include/TPCReconstruction/TPCFastTransformHelperO2.h]: ok, fixed

File: Detectors/TPC/reconstruction/include/TPCReconstruction/TPCFastTransformHelperO2.h [hunk 8]
```diff
@@ -66,9 +72,10 @@ class TPCFastTransformHelperO2
   /// initialization
   void init();
 
-  static TPCFastTransformHelperO2* sInstance; // singleton instance
+  static TPCFastTransformHelperO2* sInstance; ///< singleton instance
   bool mIsInitialized;                        ///< initialization flag
-
+  void (*mSpaceChargeCorrection)( const double XYZ[3], double dXdYdZ[3] ); ///< pointer to an external correction method
+ 
   ClassDefNV(TPCFastTransformHelperO2, 1);
```
[the remark this answers is not in the document]
Comment (sgorbuno) [reply Detectors/TPC/reconstruction/include/TPCReconstruction/TPCFastTransformHelperO2.h]: ok, fixed

File: Detectors/TPC/reconstruction/macro/getTPCTransformationExample.C (hunk 1 above)
Comment (sgorbuno) [reply Detectors/TPC/reconstruction/macro/getTPCTransformationExample.C]: I have put comments there. 
I compare the original correction to the difference between (transform with correction - transform without correction). It should be the same within the roundoff errors.

File: Detectors/TPC/reconstruction/macro/getTPCTransformationExample.C [hunk 9]
```diff
@@ -0,0 +1,105 @@
...
+  Load the macro:
+  gSystem->Load("libO2TPCFastTransformation");
+  gSystem->Load("libTPCReconstruction");
+  .x getTPCTransformationExample.C
+*/
+#if !defined(__CLING__) || defined(__ROOTCLING__)
+
+#include "TPCReconstruction/TPCFastTransformHelperO2.h"
+#include "TH1F.h"
+
```
Comment (matthiasrichter) [inline Detectors/TPC/reconstruction/macro/getTPCTransformationExample.C]:
```suggestion
#include "TStyle.h"

```

**Comment by davidrohr** [comment]
AliTPCCommon has been merged, so we can go ahead and rebase this PR and finally merge it.

**Comment by sgorbuno** [comment]
Hi @matthiasrichter,
I did all the changes you proposed; also the PR seems to compile now :)  
Can we merge it now? Or should I maybe squash all the commits to a new PR?

**Comment by davidrohr** [comment]
You dont need to squash to a new PR. We can squash during merging,or you can squash locally and then force push to this PR. This will keep the discussion history.

**Comment by matthiasrichter** [comment]
Hi Sergey, very good. There is one tiny formatting issue I want to ask you to apply the clang formatting suggestion.
```
--- a/Detectors/TPC/reconstruction/macro/getTPCTransformationExample.C
+++ b/Detectors/TPC/reconstruction/macro/getTPCTransformationExample.C
@@ -57,11 +57,11 @@ void getTPCTransformationExample()
         for (float time = 0; time < 1000; time += 30) {
 
           fastTransform->setApplyDistortionFlag(0);
-          float x0=0., y0=0., z0=0.;
+          float x0 = 0., y0 = 0., z0 = 0.;
           int err0 = fastTransform->Transform(slice, row, pad, time, x0, y0, z0);
 
           fastTransform->setApplyDistortionFlag(1);
-          float x1=0., y1=0., z1=0.;
+          float x1 = 0., y1 = 0., z1 = 0.;
           int err1 = fastTransform->Transform(slice, row, pad, time, x1, y1, z1);
 ```

You can either squash all the commits and force-push to the branch or we squash when merging. You don't need to open a new PR.

**Pull request merged**